\documentclass{article}
%\VignetteIndexEntry{Using SPAC}
%\VignetteEngine{knitr::knitr}

\usepackage[svgnames]{xcolor}

\usepackage[]{listings}
\lstloadlanguages{bash,R}
\lstset{
  tabsize=4,
  rulecolor=,
  language=R,
  basicstyle=\small\ttfamily,
  columns=fixed,
  showstringspaces=false,
  extendedchars=true,
  breaklines=true,
  breakatwhitespace,
  prebreak = \raisebox{0ex}[0ex][0ex]{\ensuremath{\hookleftarrow}},
  showtabs=false,
  showspaces=false,
  showstringspaces=false,
  keywordstyle=\color[rgb]{0.737,0.353,0.396},
  commentstyle=\color[rgb]{0.133,0.545,0.133},
  stringstyle=\color[rgb]{0.627,0.126,0.941},
  backgroundcolor=\color[rgb]{0.97,0.97,0.97},
}
\lstMakeShortInline{|}


\usepackage{hyperref}
\hypersetup{%
  linktocpage=false, % If true the page numbers in the toc are links
                     % instead of the section headings.
  pdfstartview=FitH,%
  breaklinks=true, pageanchor=true, %
  pdfpagemode=UseOutlines, plainpages=false, bookmarksnumbered, %
  bookmarksopen=true, bookmarksopenlevel=1, hypertexnames=true, %
  pdfhighlight=/O, %
  pdfauthor={\textcopyright\ K.~Oualkacha}, %
  colorlinks=true, %
  urlcolor=SteelBlue, linkcolor=blue, citecolor=LimeGreen, %
}

\title{A flexible copula-based approach for the analysis of secondary phenotypes in ascertained samples\\ \vspace{.5cm}
Working with SPAC}
\author{Karim Oualkacha, Genevi\`eve Lefebvre, Fod\'e Tounkara \& Celia M.T. Greenwood}

\begin{document}

<<include=FALSE>>=
library(knitr)
opts_chunk$set(
concordance=TRUE
)
@

\maketitle
\tableofcontents

\section{Introduction}
\label{sec:introduction}

This is a unified copula-based framework which tests for secondary phenotypes association and a single SNP in presence of selection bias. The method fits both retrospective and prospective likelihoods to handel selection bias under case-control (CC), extreme-trait (ET) and mulitple-trait (MT) sampling designs. The dependence between primary and secondary phenotypes is modelled via copulas. Thus, SPAC is a robust method for non-normality assumption of the secondary trait.

SPAC is an R package that contains methods to perform genetic association of a quantitative secondary phenotype and a single SNP in presence of selection bias. SPAC is a unified copula-based framework  fits both retrospective and prospective likelihoods to handel selection bias under case-control (CC), extreme-trait (ET) and multiple-trait (MT) sampling designs. The dependence between primary and secondary phenotypes is modelled via copulas. Thus, SPAC is a robust method for non-normality assumption of the secondary trait; improved power is possible by appropriate modelling of the primary-secondary phenotypes joint distribution via copula models.

The main user-visible function of the package is |SPAC()| function which can be used to
analyse single-SNP/Secondary phenotype association in one go.

The main function allows for prospective and retrospective copula-based likelihoods to handle selection bias via the argument \emph{method}:
  \begin{itemize}
  \item pros (default), prospective copula-based likelihood
  \item retros, retrospective copula-based likelihood
  \end{itemize}

Of note, the retrospective method does not handle covariates while the prospective-based method controls for covariates. The retrospective-based method, \emph{method="retros"}, is the computationally fastest method, and these $p$-values can be used to triage which SNPs of the genome should be re-analyzed with \emph{method="pros"}.


SPAC allows also for primary-secondary dependence modelling using five copulas

\begin{itemize}
  \item Gaussian, (default);
  \item Student, Student copula with degree of freedom equals 10;
  \item Clayton, Clayton copula;
  \item Gumbel, Gumbel copula;
  \item Frank, Frank copula.
  \end{itemize}

In order to run any of the examples below, the package needs to be installed and loaded first, of course:
<<load.pkg>>=
library(devtools)
#devtools::install_github('micau80/SPAC', ref = "micau80-patch-1", dep =FALSE)
                         #build_vignettes = TRUE, dep =FALSE)
devtools::install_github('micau80/SPAC', dep =FALSE)


library(SPAC)

@

\section{Loading the input data}
\label{sec:loading-input-data}

Before running an association test with one (or more) of the methods,
the following data needs to be present\footnote{The example code in
  this vignette uses files that are included in the SPAC
  package, that is why the \lstinline{package} argument to the
  \lstinline{system.file()} function is used.}:
\begin{itemize}
\item \emph{Phenotype data}; primary and secondary phenpotypes data should be present separately in the form of
  an R vector (one value for each individual). It is up to you (as user)
  to create the vector, for example by reading it from a CSV file using R's
  |read.csv()| function or like this:
<<load.pheno>>=
  data(data, package = "SPAC")
# data is .RData file contains 3 examples of data sets for CC, ET and MT designs
# y1cc: primary trait under the case-control (CC) design
  head(y1cc)
# y2cc: secondary trait under the case-control (CC) design
  head(y2cc)
@
\item \emph{Covariate data}; this data should be present in the form
  of a matrix. Like the phenotype data it us up to you to load this
  data. For example:
<<load.covar>>=
  data(data, package = "SPAC")
# data.file is .RData file contains 3 examples of data sets for CC, ET and MT designs
# matrix of two confounders/covariates: one dichotmous and one continuous covariate
  dim(cov.matCC)
  head(cov.matCC)
@
\item \emph{Genotype data}; SNP genotype data can be in the form of
  an R vector (one value for each individual).
<<set.genofile>>=
  data(data, package = "SPAC")
# data is .RData file contains 3 examples of data sets for CC, ET and MT designs
# SNP data for the CC design: a vector of legnth
  length(markerCC)
  head(markerCC)
@
\end{itemize}

\section{Analysing a single SNP}
\label{sec:analys-single-SNP}
With the phenotype data, the covariates and the genotype data loaded it is time for tests of association.

\subsection{Using SPAC for a case-control design}
\label{sec:using-SPAC-CC}
This is the simplest way to run SPAC for a single-SNP/secondary phenotype association test under the CC design:

<<>>=
SNPresults <- SPAC(y1 = y1cc,
                   y2 = y2cc,
                   G = markerCC,
                   covariates = as.matrix(cov.matCC),
                   link = "probit",
                   copfit = "Gaussian",
                   method = "pros",
                   Design = "CC",
                   prev = 0.1)
SNPresults
@

\begin{itemize}
\item Here the |prev| is a scalar between 0 and 1, specifies the primary phenotype prevalence.
It is needed for the |method = "pros"| and |Design = "CC"| or |Design = "MT"|.


\item The |link| is a character, specifies the link function to be used for modelling the marginal distribution of the binary primary phenotype for the CC and MT designs. The available link functions are |link=c("probit","logit","cloglog")|. The defaut is |link = "probit"|, which the liability latent model.

\item The |copfit| is a character that selects the copula model to use for modelling primary-secondary phenptypes dependence. Can be one of the following:
\begin{itemize}
\item |copfit = "Gaussian"|, (default)
\item |copfit = "Student"|, Student copula with degree of freedom equals 10
\item |copfit = "Clayton"|, Clayton copula
\item |copfit = "Gumbel"|, Gumbel copula
\item |copfit = "Frank"|, Frank copula
\end{itemize}
\end{itemize}

\subsection{Using SPAC for an extrem-trait (ET) design}
\label{sec:using-SPAC-ET}
The next code run SPAC for a single-SNP/secondary phenotype association test under the ET design:

<<>>=
SNPresults.ET <- SPAC(y1 = y1et,
                   y2 = y2et,
                   G = markeret,
                   covariates = as.matrix(cov.matCC),
                   copfit = "Gaussian",
                   method = "pros",
                   Design = "ET",
                   cutoffs = c(-1.498953,2.174215),
                   p.lu =c(0.1,0.1)
                   )
SNPresults.ET
@



\begin{itemize}
\item The |cutoffs| is a vector or scalar, depending on the sampling mechanism design |Design = "ET"| or  |Design = "MT"|. For the ET design the |cutoffs| is a $2\times 1$ vector, |cutoffs = c(ylb,yub)|, with
\begin{itemize}
\item ylb is the lower primary trait threshold
\item yub is the upper primary trait threshold
\end{itemize}

\item The |p.lu| is a vector specifying the proportion of individuals with lower and upper extreme primary trait. It is needed for the |method = "pros"| and |Design = "ET"|.

\end{itemize}


Of note, the ET design does not need a |link| argument since it fits a liability threshold model for the primary phenotype.

\subsection{Using SPAC for an multiple-trait (MT) design}
\label{sec:using-SPAC-MT}
The next code run SPAC for a single-SNP/secondary phenotype association test under the MT design. Here we use |copfit = "Clayton"| as a copula model to model the joint distribution of the primary-secondary phenotypes:
<<>>=
SNPresults.MT <- SPAC(y1 = y1mt,
                   y2 = y2mt,
                   G = markermt,
                   covariates = as.matrix(cov.matMT),
                   link = "probit",
                   copfit = "Clayton",
                   method = "pros",
                   Design = "MT",
                   cutoffs = 1.865465,
                   prev = 0.1,
                   prev2 =0.06112)
SNPresults.MT
@


\begin{itemize}

\item In the MT design, the |cutoffs| is a scalar, |cutoffs = y2ub|, with |y2ub| is the upper secondary trait threshold

\item The |prev2| is a scalar between 0 and 1, specifies the proportion of diseased individuals with secondary trait exceeding the threshold |y2ub|.
 It is needed for the |method ="pros"| and |Design = "MT"|.


\end{itemize}

\begin{thebibliography}{99}
\bibitem{Tounkara2019} Fod\'e Tounkara, Genevi\`eve Lefebvre, Celia MT Greenwood and Karim Oualkacha (2019). \emph{A flexible copula-based approach for the analysis of secondary phenotypes in ascertained samples}. Statistics in Medicine. 1-32. Under revision.
\end{thebibliography}

\end{document}


